File: csharp/DCbor/DCbor/Cbor.cs
using System.Globalization;
using System.Text;

namespace BlockchainCommons.DCbor;

/// <summary>
/// The central immutable type representing a CBOR data item.
/// Wraps a <see cref="CborCase"/> and provides encoding, decoding,
/// equality, formatting, and convenience methods.
/// </summary>
public sealed class Cbor : IEquatable<Cbor>
{
    private readonly CborCase _case;

    public Cbor(CborCase cborCase)
    {
        _case = cborCase;
    }

    /// <summary>Returns the underlying discriminated-union case.</summary>
    public CborCase Case => _case;

    // --- Encoding ---

    /// <summary>Encodes this CBOR value to canonical dCBOR bytes.</summary>
    public byte[] ToCborData()
    {
        switch (_case)
        {
            case CborCase.UnsignedCase u:
                return Varint.EncodeVarInt(u.Value, MajorType.Unsigned);

            case CborCase.NegativeCase n:
                return Varint.EncodeVarInt(n.Value, MajorType.Negative);

            case CborCase.ByteStringCase bs:
            {
                var header = Varint.EncodeVarInt((ulong)bs.Value.Length, MajorType.ByteString);
                var result = new byte[header.Length + bs.Value.Length];
                header.CopyTo(result, 0);
                bs.Value.DataRef.CopyTo(result, header.Length);
                return result;
            }

            case CborCase.TextCase t:
            {
                string nfc = t.Value.IsNormalized(NormalizationForm.FormC)
                    ? t.Value
                    : t.Value.Normalize(NormalizationForm.FormC);
                byte[] utf8 = Encoding.UTF8.GetBytes(nfc);
                var header = Varint.EncodeVarInt((ulong)utf8.Length, MajorType.Text);
                var result = new byte[header.Length + utf8.Length];
                header.CopyTo(result, 0);
                utf8.CopyTo(result, header.Length);
                return result;
            }

            case CborCase.ArrayCase a:
            {
                var buf = new List<byte>(Varint.EncodeVarInt((ulong)a.Value.Count, MajorType.Array));
                foreach (var item in a.Value)
                    buf.AddRange(item.ToCborData());
                return buf.ToArray();
            }

            case CborCase.MapCase m:
                return m.Value.CborData();

            case CborCase.TaggedCase tg:
            {
                var header = Varint.EncodeVarInt(tg.Tag.Value, MajorType.Tagged);
                var itemData = tg.Item.ToCborData();
                var result = new byte[header.Length + itemData.Length];
                header.CopyTo(result, 0);
                itemData.CopyTo(result, header.Length);
                return result;
            }

            case CborCase.SimpleCase s:
                return s.Value.CborData();

            default:
                throw new InvalidOperationException();
        }
    }

    /// <summary>Returns the hex-encoded CBOR bytes.</summary>
    public string Hex() => Convert.ToHexString(ToCborData()).ToLowerInvariant();

    // --- Decoding ---

    /// <summary>Decodes binary data into a CBOR value, validating dCBOR rules.</summary>
    public static Cbor TryFromData(ReadOnlySpan<byte> data)
    {
        return CborDecoder.DecodeCbor(data);
    }

    /// <summary>Decodes a hex string into a CBOR value.</summary>
    public static Cbor TryFromHex(string hex)
    {
        byte[] data = Convert.FromHexString(hex);
        return TryFromData(data);
    }

    // --- From integer types ---

    public static Cbor FromInt(long value)
    {
        if (value < 0)
        {
            ulong n = (ulong)(-1 - (Int128)value);
            return new Cbor(CborCase.Negative(n));
        }
        return new Cbor(CborCase.Unsigned((ulong)value));
    }

    public static Cbor FromUInt(ulong value)
    {
        return new Cbor(CborCase.Unsigned(value));
    }

    // --- From float ---

    public static Cbor FromDouble(double value)
    {
        // Numeric reduction: try integer
        if (value < 0.0)
        {
            var n = ExactFrom.Int128FromDouble(value);
            if (n.HasValue)
            {
                var i = ExactFrom.UInt64FromInt128(-1 - n.Value);
                if (i.HasValue)
                    return new Cbor(CborCase.Negative(i.Value));
            }
        }

        var ui = ExactFrom.UInt64FromDouble(value);
        if (ui.HasValue)
            return new Cbor(CborCase.Unsigned(ui.Value));

        return new Cbor(CborCase.FromSimple(Simple.Float(value)));
    }

    public static Cbor FromFloat(float value)
    {
        if (value < 0.0f)
        {
            var i = ExactFrom.UInt64FromFloat(-1f - value);
            if (i.HasValue)
                return new Cbor(CborCase.Negative(i.Value));
        }

        var ui = ExactFrom.UInt32FromFloat(value);
        if (ui.HasValue)
            return new Cbor(CborCase.Unsigned(ui.Value));

        return new Cbor(CborCase.FromSimple(Simple.Float(value)));
    }

    // --- From other types ---

    public static Cbor FromString(string value) => new(CborCase.Text(value));
    public static Cbor FromBool(bool value) => new(CborCase.FromSimple(value ? Simple.True : Simple.False));
    public static Cbor FromByteString(byte[] data) => new(CborCase.FromByteString(new ByteString(data)));
    public static Cbor FromByteString(ByteString value) => new(CborCase.FromByteString(value));

    // --- Collection conversions ---

    /// <summary>Creates a CBOR array from a list of values.</summary>
    public static Cbor FromList(IEnumerable<Cbor> items)
    {
        return new Cbor(CborCase.Array(items.ToList()));
    }

    /// <summary>Creates a CBOR array from a list of integers.</summary>
    public static Cbor FromIntList(IEnumerable<int> items)
    {
        return FromList(items.Select(i => FromInt(i)));
    }

    /// <summary>Creates a CBOR array from a list of strings.</summary>
    public static Cbor FromStringList(IEnumerable<string> items)
    {
        return FromList(items.Select(s => FromString(s)));
    }

    /// <summary>Creates a CBOR map from a dictionary of int keys and string values.</summary>
    public static Cbor FromDictionary(IDictionary<int, string> dict)
    {
        var m = new CborMap();
        foreach (var kv in dict)
            m.Insert(FromInt(kv.Key), FromString(kv.Value));
        return new Cbor(CborCase.Map(m));
    }

    /// <summary>Extracts an array of typed values from a CBOR array.</summary>
    public List<int> TryIntoIntList()
    {
        var arr = TryIntoArray();
        return arr.Select(c => c.TryIntoInt32()).ToList();
    }

    /// <summary>Extracts a Dictionary from a CBOR map.</summary>
    public Dictionary<int, string> TryIntoDictionaryIntString()
    {
        var map = TryIntoMap();
        var result = new Dictionary<int, string>();
        foreach (var (key, value) in map)
            result[key.TryIntoInt32()] = value.TryIntoText();
        return result;
    }

    /// <summary>Extracts a SortedDictionary from a CBOR map.</summary>
    public SortedDictionary<int, string> TryIntoSortedDictionaryIntString()
    {
        var map = TryIntoMap();
        var result = new SortedDictionary<int, string>();
        foreach (var (key, value) in map)
            result[key.TryIntoInt32()] = value.TryIntoText();
        return result;
    }

    /// <summary>Extracts a HashSet of ints from a CBOR array.</summary>
    public HashSet<int> TryIntoHashSetInt()
    {
        var arr = TryIntoArray();
        return arr.Select(c => c.TryIntoInt32()).ToHashSet();
    }

    // --- Convenience factories ---

    public static Cbor Null() => new(CborCase.FromSimple(Simple.Null));
    public static Cbor True() => new(CborCase.FromSimple(Simple.True));
    public static Cbor False() => new(CborCase.FromSimple(Simple.False));
    public static Cbor Nan() => new(CborCase.FromSimple(Simple.Float(double.NaN)));

    public static Cbor ToByteString(byte[] data) => FromByteString(data);
    public static Cbor ToByteString(ReadOnlySpan<byte> data) => new(CborCase.FromByteString(new ByteString(data)));

    public static Cbor ToByteStringFromHex(string hex)
    {
        byte[] data = Convert.FromHexString(hex);
        return FromByteString(data);
    }

    public static Cbor ToTaggedValue(ulong tag, Cbor item)
    {
        return new Cbor(CborCase.Tagged(new Tag(tag), item));
    }

    public static Cbor ToTaggedValue(Tag tag, Cbor item)
    {
        return new Cbor(CborCase.Tagged(tag, item));
    }

    // --- Type inspection ---

    public bool IsNull => _case is CborCase.SimpleCase s && s.Value is Simple.NullValue;
    public bool IsTrue => _case is CborCase.SimpleCase s && s.Value is Simple.TrueValue;
    public bool IsFalse => _case is CborCase.SimpleCase s && s.Value is Simple.FalseValue;
    public bool IsBool => _case is CborCase.SimpleCase s && (s.Value is Simple.TrueValue || s.Value is Simple.FalseValue);
    public bool IsNumber => _case is CborCase.UnsignedCase or CborCase.NegativeCase or CborCase.SimpleCase { Value: Simple.FloatValue };
    public bool IsNan => _case is CborCase.SimpleCase s && s.Value.IsNaN;
    public bool IsText => _case is CborCase.TextCase;
    public bool IsByteString => _case is CborCase.ByteStringCase;
    public bool IsArray => _case is CborCase.ArrayCase;
    public bool IsMap => _case is CborCase.MapCase;
    public bool IsTaggedValue => _case is CborCase.TaggedCase;

    // --- Type extraction ---

    public bool? AsBool()
    {
        return _case switch
        {
            CborCase.SimpleCase { Value: Simple.TrueValue } => true,
            CborCase.SimpleCase { Value: Simple.FalseValue } => false,
            _ => null,
        };
    }

    public string? AsText()
    {
        return _case is CborCase.TextCase t ? t.Value : null;
    }

    public byte[]? AsByteStringData()
    {
        return _case is CborCase.ByteStringCase bs ? bs.Value.ToArray() : null;
    }

    public IReadOnlyList<Cbor>? AsArray()
    {
        return _case is CborCase.ArrayCase a ? a.Value : null;
    }

    public CborMap? AsMap()
    {
        return _case is CborCase.MapCase m ? m.Value : null;
    }

    public (Tag Tag, Cbor Item)? AsTaggedValue()
    {
        return _case is CborCase.TaggedCase tg ? (tg.Tag, tg.Item) : null;
    }

    // --- Try extraction (throwing) ---

    public bool TryIntoBool()
    {
        return AsBool() ?? throw new CborWrongTypeException();
    }

    public string TryIntoText()
    {
        return AsText() ?? throw new CborWrongTypeException();
    }

    public byte[] TryIntoByteString()
    {
        return AsByteStringData() ?? throw new CborWrongTypeException();
    }

    public IReadOnlyList<Cbor> TryIntoArray()
    {
        return AsArray() ?? throw new CborWrongTypeException();
    }

    public CborMap TryIntoMap()
    {
        return AsMap() ?? throw new CborWrongTypeException();
    }

    public (Tag Tag, Cbor Item) TryIntoTaggedValue()
    {
        return AsTaggedValue() ?? throw new CborWrongTypeException();
    }

    public Cbor TryIntoExpectedTaggedValue(Tag expectedTag)
    {
        var (tag, item) = TryIntoTaggedValue();
        if (tag != expectedTag)
            throw new CborWrongTagException(expectedTag, tag);
        return item;
    }

    public Cbor TryIntoExpectedTaggedValue(ulong expectedTagValue)
    {
        return TryIntoExpectedTaggedValue(new Tag(expectedTagValue));
    }

    // --- Numeric extraction ---

    public ulong TryIntoUInt64()
    {
        return _case switch
        {
            CborCase.UnsignedCase u => u.Value,
            _ => throw new CborWrongTypeException(),
        };
    }

    public long TryIntoInt64()
    {
        return _case switch
        {
            CborCase.UnsignedCase u when u.Value <= (ulong)long.MaxValue => (long)u.Value,
            CborCase.NegativeCase n when n.Value <= (ulong)long.MaxValue => -1L - (long)n.Value,
            CborCase.UnsignedCase or CborCase.NegativeCase => throw new CborOutOfRangeException(),
            _ => throw new CborWrongTypeException(),
        };
    }

    public int TryIntoInt32()
    {
        long v = TryIntoInt64();
        if (v < int.MinValue || v > int.MaxValue)
            throw new CborOutOfRangeException();
        return (int)v;
    }

    public double TryIntoDouble()
    {
        return _case switch
        {
            CborCase.UnsignedCase u =>
                ExactFrom.DoubleFromUInt64(u.Value) ?? throw new CborOutOfRangeException(),
            CborCase.NegativeCase n =>
                ExactFrom.DoubleFromUInt64(n.Value) is { } f
                    ? -1.0 - f
                    : throw new CborOutOfRangeException(),
            CborCase.SimpleCase { Value: Simple.FloatValue fv } => fv.Value,
            _ => throw new CborWrongTypeException(),
        };
    }

    // --- Equality ---

    public bool Equals(Cbor? other)
    {
        if (other is null) return false;
        return _case.Equals(other._case);
    }

    public override bool Equals(object? obj) => Equals(obj as Cbor);

    public override int GetHashCode() => _case.GetHashCode();

    public static bool operator ==(Cbor? a, Cbor? b)
    {
        if (a is null) return b is null;
        return a.Equals(b);
    }

    public static bool operator !=(Cbor? a, Cbor? b) => !(a == b);

    // --- Display ---

    public override string ToString()
    {
        return _case switch
        {
            CborCase.UnsignedCase u => u.Value.ToString(),
            CborCase.NegativeCase n => (-1 - (Int128)n.Value).ToString(),
            CborCase.ByteStringCase bs => $"h'{bs.Value.ToHex()}'",
            CborCase.TextCase t => FormatString(t.Value),
            CborCase.ArrayCase a => FormatArray(a.Value),
            CborCase.MapCase m => m.Value.ToString()!,
            CborCase.TaggedCase tg => $"{tg.Tag}({tg.Item})",
            CborCase.SimpleCase s => s.Value.ToString()!,
            _ => throw new InvalidOperationException(),
        };
    }

    /// <summary>
    /// A verbose debug representation showing the internal structure.
    /// </summary>
    public string DebugDescription
    {
        get
        {
            return _case switch
            {
                CborCase.UnsignedCase u => $"unsigned({u.Value})",
                CborCase.NegativeCase n => $"negative({-1 - (Int128)n.Value})",
                CborCase.ByteStringCase bs => $"bytes({bs.Value.ToHex()})",
                CborCase.TextCase t => $"text(\"{t.Value}\")",
                CborCase.ArrayCase a => $"array([{string.Join(", ", a.Value.Select(x => x.DebugDescription))}])",
                CborCase.MapCase m => FormatMapDebug(m.Value),
                CborCase.TaggedCase tg => $"tagged({tg.Tag}, {tg.Item.DebugDescription})",
                CborCase.SimpleCase s => $"simple({s.Value.Name()})",
                _ => throw new InvalidOperationException(),
            };
        }
    }

    private static string FormatString(string s)
    {
        var sb = new StringBuilder();
        sb.Append('"');
        foreach (char c in s)
        {
            if (c == '"')
                sb.Append("\\\"");
            else
                sb.Append(c);
        }
        sb.Append('"');
        return sb.ToString();
    }

    private static string FormatArray(IReadOnlyList<Cbor> a)
    {
        return "[" + string.Join(", ", a.Select(x => x.ToString())) + "]";
    }

    private static string FormatMapDebug(CborMap m)
    {
        var items = new List<string>();
        foreach (var (key, value) in m)
        {
            string keyHex = "0x" + Convert.ToHexString(key.ToCborData()).ToLowerInvariant();
            items.Add($"{keyHex}: ({key.DebugDescription}, {value.DebugDescription})");
        }
        return "map({" + string.Join(", ", items) + "})";
    }
}

File: python/dcbor/src/dcbor/cbor.py
from __future__ import annotations

import math
from enum import IntEnum
from typing import TYPE_CHECKING, Any

from .byte_string import ByteString
from .error import (
    DCBORError,
    MissingMapKey,
    OutOfRange,
    WrongTag,
    WrongType,
)
from .float_utils import _exact_neg_u64_from_float, _exact_uint_from_float
from .simple import Simple
from .string_util import to_nfc
from .tag import Tag
from .varint import MajorType, encode_varint

if TYPE_CHECKING:
    from .map import Map

_U64_MAX = 0xFFFFFFFFFFFFFFFF


class CBORCase(IntEnum):
    """Discriminant enum identifying which CBOR major type a CBOR value holds."""

    UNSIGNED = 0
    NEGATIVE = 1
    BYTE_STRING = 2
    TEXT = 3
    ARRAY = 4
    MAP = 5
    TAGGED = 6
    SIMPLE = 7


class CBOR:
    """Deterministic CBOR value supporting all standard major types."""

    __slots__ = ("_case", "_value")

    def __init__(self, case: CBORCase, value: Any) -> None:
        self._case = case
        self._value = value

    # --- Case accessors ---

    @property
    def case(self) -> CBORCase:
        return self._case

    @property
    def value(self) -> Any:
        return self._value

    # --- Factory methods for construction ---

    @staticmethod
    def from_int(value: int) -> CBOR:
        """Create a CBOR unsigned or negative integer from a Python int."""
        if value < 0:
            n = -1 - value
            if n < 0 or n > _U64_MAX:
                raise ValueError(f"integer {value} out of CBOR range")
            return CBOR(CBORCase.NEGATIVE, n)
        else:
            if value > _U64_MAX:
                raise ValueError(f"integer {value} out of CBOR range")
            return CBOR(CBORCase.UNSIGNED, value)

    @staticmethod
    def from_float(value: float) -> CBOR:
        """Create a CBOR value from a float, preferring integer encoding when lossless."""
        if value < 0.0:
            neg = _exact_neg_u64_from_float(value)
            if neg is not None:
                return CBOR(CBORCase.NEGATIVE, neg)
        pos = _exact_uint_from_float(value, _U64_MAX)
        if pos is not None:
            return CBOR(CBORCase.UNSIGNED, pos)
        return CBOR(CBORCase.SIMPLE, Simple.make_float(value))

    @staticmethod
    def from_bool(value: bool) -> CBOR:
        """Create a CBOR simple true or false value."""
        if value:
            return CBOR(CBORCase.SIMPLE, Simple.make_true())
        return CBOR(CBORCase.SIMPLE, Simple.make_false())

    @staticmethod
    def from_bytes(data: bytes | bytearray | ByteString) -> CBOR:
        """Create a CBOR byte string from bytes, bytearray, or ByteString."""
        if isinstance(data, ByteString):
            return CBOR(CBORCase.BYTE_STRING, data)
        return CBOR(CBORCase.BYTE_STRING, ByteString(data))

    @staticmethod
    def from_text(value: str) -> CBOR:
        """Create a CBOR text string from a Python str."""
        return CBOR(CBORCase.TEXT, value)

    @staticmethod
    def from_array(items: list[CBOR]) -> CBOR:
        """Create a CBOR array from a list of CBOR values."""
        return CBOR(CBORCase.ARRAY, list(items))

    @staticmethod
    def from_map(m: Map) -> CBOR:
        """Create a CBOR map value wrapping a Map instance."""
        return CBOR(CBORCase.MAP, m)

    @staticmethod
    def from_tagged_value(tag: Tag | int, item: CBOR) -> CBOR:
        """Create a CBOR tagged value from a tag (or raw tag number) and a CBOR item."""
        if isinstance(tag, int):
            tag = Tag(tag)
        return CBOR(CBORCase.TAGGED, (tag, item))

    @staticmethod
    def null() -> CBOR:
        return CBOR(CBORCase.SIMPLE, Simple.make_null())

    @staticmethod
    def cbor_true() -> CBOR:
        return CBOR(CBORCase.SIMPLE, Simple.make_true())

    @staticmethod
    def cbor_false() -> CBOR:
        return CBOR(CBORCase.SIMPLE, Simple.make_false())

    @staticmethod
    def nan() -> CBOR:
        return CBOR(CBORCase.SIMPLE, Simple.make_float(float("nan")))

    @staticmethod
    def from_value(value: int | float | str | bool | bytes | bytearray | ByteString | CBOR | list | None) -> CBOR:
        """Convert a native Python value to CBOR; accepts int, float, str, bool, bytes, list, or CBOR."""
        if isinstance(value, CBOR):
            return value
        if isinstance(value, bool):
            return CBOR.from_bool(value)
        if isinstance(value, int):
            return CBOR.from_int(value)
        if isinstance(value, float):
            return CBOR.from_float(value)
        if isinstance(value, str):
            return CBOR.from_text(value)
        if isinstance(value, (bytes, bytearray, ByteString)):
            return CBOR.from_bytes(value)
        if isinstance(value, Tag):
            raise TypeError("Use CBOR.from_tagged_value(tag, item) for tagged values")
        if isinstance(value, list):
            return CBOR.from_array([CBOR.from_value(x) for x in value])
        raise TypeError(f"Cannot convert {type(value).__name__} to CBOR")

    # --- Encoding ---

    def to_cbor_data(self) -> bytes:
        """Encode this value to deterministic CBOR bytes."""
        match self._case:
            case CBORCase.UNSIGNED:
                return encode_varint(self._value, MajorType.UNSIGNED)
            case CBORCase.NEGATIVE:
                return encode_varint(self._value, MajorType.NEGATIVE)
            case CBORCase.BYTE_STRING:
                data = self._value.data
                return encode_varint(len(data), MajorType.BYTE_STRING) + data
            case CBORCase.TEXT:
                text = to_nfc(self._value)
                encoded = text.encode("utf-8")
                return encode_varint(len(encoded), MajorType.TEXT) + encoded
            case CBORCase.ARRAY:
                result = encode_varint(len(self._value), MajorType.ARRAY)
                for item in self._value:
                    result += item.to_cbor_data()
                return result
            case CBORCase.MAP:
                return self._value.cbor_data()
            case CBORCase.TAGGED:
                tag, item = self._value
                return (
                    encode_varint(tag.value, MajorType.TAGGED)
                    + item.to_cbor_data()
                )
            case CBORCase.SIMPLE:
                return self._value.cbor_data()

    @staticmethod
    def from_data(data: bytes | bytearray) -> CBOR:
        """Decode a CBOR value from raw bytes, raising on any non-canonical encoding."""
        from .decode import decode_cbor
        return decode_cbor(data)

    @staticmethod
    def from_hex(hex_str: str) -> CBOR:
        """Decode a CBOR value from a hexadecimal string."""
        return CBOR.from_data(bytes.fromhex(hex_str))

    # --- Type checks ---

    @property
    def is_unsigned(self) -> bool:
        return self._case == CBORCase.UNSIGNED

    @property
    def is_negative(self) -> bool:
        return self._case == CBORCase.NEGATIVE

    @property
    def is_byte_string(self) -> bool:
        return self._case == CBORCase.BYTE_STRING

    @property
    def is_text(self) -> bool:
        return self._case == CBORCase.TEXT

    @property
    def is_array(self) -> bool:
        return self._case == CBORCase.ARRAY

    @property
    def is_map(self) -> bool:
        return self._case == CBORCase.MAP

    @property
    def is_tagged_value(self) -> bool:
        return self._case == CBORCase.TAGGED

    @property
    def is_simple(self) -> bool:
        return self._case == CBORCase.SIMPLE

    @property
    def is_bool(self) -> bool:
        if self._case != CBORCase.SIMPLE:
            return False
        s: Simple = self._value
        return s.is_true or s.is_false

    @property
    def is_true(self) -> bool:
        return self._case == CBORCase.SIMPLE and self._value.is_true

    @property
    def is_false(self) -> bool:
        return self._case == CBORCase.SIMPLE and self._value.is_false

    @property
    def is_null(self) -> bool:
        return self._case == CBORCase.SIMPLE and self._value.is_null

    @property
    def is_nan(self) -> bool:
        return self._case == CBORCase.SIMPLE and self._value.is_nan

    @property
    def is_float(self) -> bool:
        return self._case == CBORCase.SIMPLE and self._value.is_float

    @property
    def is_number(self) -> bool:
        return self._case in (
            CBORCase.UNSIGNED,
            CBORCase.NEGATIVE,
        ) or self.is_float

    # --- Typed extractors ---

    def try_int(self) -> int:
        if self._case == CBORCase.UNSIGNED:
            return self._value
        if self._case == CBORCase.NEGATIVE:
            return -1 - self._value
        raise WrongType()

    def try_float(self) -> float:
        if self._case == CBORCase.UNSIGNED:
            return float(self._value)
        if self._case == CBORCase.NEGATIVE:
            return float(-1 - self._value)
        if self._case == CBORCase.SIMPLE and self._value.is_float:
            return self._value.float_value
        raise WrongType()

    def try_bool(self) -> bool:
        if self._case == CBORCase.SIMPLE:
            if self._value.is_true:
                return True
            if self._value.is_false:
                return False
        raise WrongType()

    def as_bool(self) -> bool | None:
        try:
            return self.try_bool()
        except WrongType:
            return None

    def try_byte_string(self) -> bytes:
        if self._case == CBORCase.BYTE_STRING:
            return self._value.data
        raise WrongType()

    def as_byte_string(self) -> bytes | None:
        if self._case == CBORCase.BYTE_STRING:
            return self._value.data
        return None

    def try_text(self) -> str:
        if self._case == CBORCase.TEXT:
            return self._value
        raise WrongType()

    def as_text(self) -> str | None:
        if self._case == CBORCase.TEXT:
            return self._value
        return None

    def try_array(self) -> list[CBOR]:
        if self._case == CBORCase.ARRAY:
            return list(self._value)
        raise WrongType()

    def as_array(self) -> list[CBOR] | None:
        if self._case == CBORCase.ARRAY:
            return list(self._value)
        return None

    def try_map(self) -> Map:
        from .map import Map
        if self._case == CBORCase.MAP:
            return self._value
        raise WrongType()

    def as_map(self) -> Map | None:
        if self._case == CBORCase.MAP:
            return self._value
        return None

    def try_tagged_value(self) -> tuple[Tag, CBOR]:
        if self._case == CBORCase.TAGGED:
            return self._value
        raise WrongType()

    def as_tagged_value(self) -> tuple[Tag, CBOR] | None:
        if self._case == CBORCase.TAGGED:
            return self._value
        return None

    def try_expected_tagged_value(self, expected_tag: Tag | int) -> CBOR:
        if isinstance(expected_tag, int):
            expected_tag = Tag(expected_tag)
        if self._case == CBORCase.TAGGED:
            tag, item = self._value
            if tag == expected_tag:
                return item
            raise WrongTag(expected_tag, tag)
        raise WrongType()

    def try_simple(self) -> Simple:
        if self._case == CBORCase.SIMPLE:
            return self._value
        raise WrongType()

    # --- Diagnostics (delegate to diag module) ---

    def diagnostic(self) -> str:
        from .diag import diagnostic_impl
        return diagnostic_impl(self, annotate=False, flat=False, summarize=False)

    def diagnostic_annotated(self, tags_store: object = None) -> str:
        from .diag import diagnostic_impl
        return diagnostic_impl(
            self, annotate=True, flat=False, summarize=False,
            tags_store=tags_store,
        )

    def diagnostic_flat(self) -> str:
        from .diag import diagnostic_impl
        return diagnostic_impl(self, annotate=False, flat=True, summarize=False)

    def summary(self, tags_store: object = None) -> str:
        from .diag import diagnostic_impl
        return diagnostic_impl(
            self, annotate=True, flat=True, summarize=True,
            tags_store=tags_store,
        )

    # --- Hex dump (delegate to dump module) ---

    def hex(self) -> str:
        """Return the deterministic CBOR encoding of this value as a hex string."""
        return self.to_cbor_data().hex()

    def hex_annotated(self, tags_store: object = None) -> str:
        from .dump import hex_annotated_impl
        return hex_annotated_impl(self, tags_store=tags_store)

    # --- Walk (delegate to walk module) ---

    def walk(self, state: object, visitor: object) -> object:
        from .walk import walk_impl
        return walk_impl(self, state, visitor)

    # --- Equality and hashing ---

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, CBOR):
            return NotImplemented
        if self._case != other._case:
            return False
        if self._case == CBORCase.TAGGED:
            t1, v1 = self._value
            t2, v2 = other._value
            return t1 == t2 and v1 == v2
        return self._value == other._value

    def __hash__(self) -> int:
        match self._case:
            case CBORCase.UNSIGNED | CBORCase.NEGATIVE | CBORCase.TEXT:
                return hash((self._case, self._value))
            case CBORCase.BYTE_STRING:
                return hash((self._case, self._value.data))
            case CBORCase.ARRAY:
                return hash((self._case, tuple(self._value)))
            case CBORCase.MAP:
                return hash((self._case, self._value))
            case CBORCase.TAGGED:
                tag, item = self._value
                return hash((self._case, tag, item))
            case CBORCase.SIMPLE:
                return hash((self._case, self._value))

    # --- Leaf formatting (avoids circular call through diag module) ---

    def _leaf_str(self) -> str | None:
        """Direct string for leaf CBOR types (used by diagnostic and __str__)."""
        match self._case:
            case CBORCase.UNSIGNED:
                return str(self._value)
            case CBORCase.NEGATIVE:
                return str(-1 - self._value)
            case CBORCase.BYTE_STRING:
                return f"h'{self._value.data.hex()}'"
            case CBORCase.TEXT:
                return f'"{self._value}"'
            case CBORCase.SIMPLE:
                return str(self._value)
        return None

    def debug_description(self) -> str:
        """Detailed debug representation showing the CBOR structure."""
        match self._case:
            case CBORCase.UNSIGNED:
                return f"unsigned({self._value})"
            case CBORCase.NEGATIVE:
                return f"negative({-1 - self._value})"
            case CBORCase.BYTE_STRING:
                return f"bytes({self._value.data.hex()})"
            case CBORCase.TEXT:
                return f'text("{self._value}")'
            case CBORCase.ARRAY:
                items = ", ".join(item.debug_description() for item in self._value)
                return f"array([{items}])"
            case CBORCase.MAP:
                pairs = []
                for key, val in self._value.iter():
                    key_hex = "0x" + key.to_cbor_data().hex()
                    pairs.append(
                        f"{key_hex}: ({key.debug_description()}, {val.debug_description()})"
                    )
                return "map({" + ", ".join(pairs) + "})"
            case CBORCase.TAGGED:
                tag, item = self._value
                return f"tagged({tag}, {item.debug_description()})"
            case CBORCase.SIMPLE:
                return f"simple({self._value!r})"

    def __repr__(self) -> str:
        return self.debug_description()

    def __str__(self) -> str:
        leaf = self._leaf_str()
        if leaf is not None:
            return leaf
        match self._case:
            case CBORCase.ARRAY:
                return "[" + ", ".join(str(item) for item in self._value) + "]"
            case CBORCase.MAP:
                pairs = []
                for key, val in self._value.iter():
                    pairs.append(f"{key}: {val}")
                return "{" + ", ".join(pairs) + "}"
            case CBORCase.TAGGED:
                tag, item = self._value
                if tag.name is not None:
                    return f"{tag.name}({item})"
                return f"{tag.value}({item})"
